Allow ADNodalBC to process non-local dofs

lindsayad opened the pull request:
Closes #23940

lindsayad committed:
Allow ADNodalBC to process non-local dofs

Closes #23940

lindsayad committed:
Update class description for NeumannBC

lindsayad committed:
Try reducing calls to close

lindsayad committed:
Test nonlocal dof dependence (#23940)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1440917/) on 2407e9a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23941/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/1440913/) on 2407e9a : invalidated by @lindsayad

moosebuild commented:
Job [No optional](https://civet.inl.gov/job/1440928/) on 2407e9a : invalidated by @lindsayad

moosebuild commented:
Job [Python 3.8](https://civet.inl.gov/job/1440902/) on 2407e9a : invalidated by @lindsayad

moosebuild commented:
Job [Python 3.9](https://civet.inl.gov/job/1440903/) on 2407e9a : invalidated by @lindsayad

moosebuild commented:
Job [Conda build (ARM Mac)](https://civet.inl.gov/job/1447931/) on 8d214cd : invalidated by @lindsayad

moosebuild commented:
Job [Conda build (Linux)](https://civet.inl.gov/job/1447933/) on 8d214cd : invalidated by @lindsayad

moosebuild commented:
Job [Conda build (Intel Mac)](https://civet.inl.gov/job/1447932/) on 8d214cd : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1447965/) on 8d214cd wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>5f4c09</th>
      <th colspan=4>#23941 8d214c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.96%</td>
      <td>84.96%</td>
      <td>-0.00%</td>
      <td>97.47%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>85099</td>
      <td>85082</td>
      <td>-17</td>
      <td>77</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15066</td>
      <td>15067</td>
      <td>+1</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23941/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23941/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23941/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on test/src/relationshipmanagers/GhostEverything.C:
```diff
@@ -0,0 +1,65 @@
+std::string
+GhostEverything::getInfo() const
+{
+  std::ostringstream oss;
+  oss << "GhostEverything";
+  return oss.str();
```
```suggestion
  return "GhostEverything";
```

GiudGiud commented on test/tests/bcs/ad_nodal_bc_nonlocal_dependence/test.i:
```diff
@@ -0,0 +1,60 @@
+    variable = u
+  []
+  [pin]
+    type = ADAverageValuePin
+    variable = u
+    boundary = 'pin'
```
```suggestion
    type = ADAverageValuePin
    variable = u
    # This BC actually constrains all the domain, not just the boundary
    boundary = 'pin'
```

lindsayad replied:
It really does only constrain the single degree of freedom on the boundary `pin` in order to remove this problem's nullspace. But by constraining the single dof it does ensure that the average value in the entire domain is 0
```suggestion
    type = ADAverageValuePin
    variable = u
    # This BC constrains a single degree of freedom on the node specified by the boundary name 'pin' but since 
    # only gradients of 'u' appear in the weak form, constraining the single degree of freedom will make it appear as 
    # if we have constrained all the degrees of freedom in the system such that we have an average value of 0. 
     boundary = 'pin'
```

GiudGiud commented on framework/src/bcs/ADNodalBC.C:
```diff
@@ -44,7 +44,8 @@ ADNodalBCTempl<T>::ADNodalBCTempl(const InputParameters & parameters)
     _current_node(_var.node()),
     _u(_var.adNodalValue()),
     _set_components(
-        {getParam<bool>("set_x_comp"), getParam<bool>("set_y_comp"), getParam<bool>("set_z_comp")})
+        {getParam<bool>("set_x_comp"), getParam<bool>("set_y_comp"), getParam<bool>("set_z_comp")}),
```
ugh we should suppress those for normal variables??

lindsayad replied:
removed them for non-vector instantiations

GiudGiud commented on framework/src/bcs/ADNodalBC.C:
```diff
@@ -67,148 +68,92 @@ conversionHelper(const libMesh::VectorValue<ADReal> & value, const unsigned int
 void
 ADNodalBCTempl<T>::computeOffDiagJacobian(const unsigned int jvar_num)
 {
+  // Only need to do this once because AD does all the derivatives at once
   if (jvar_num == _var.number())
     computeJacobian();
```
are we worried it will double count then?
like a call to computeJacobian will have already happened for this variable?

lindsayad replied:
`NonlinearSystemBase::computeJacobianInternal` only makes calls to `computeOffDiagJacobian` for all coupling entries including ivar == jvar

GiudGiud commented on framework/src/bcs/ADNodalBC.C:
```diff
@@ -67,148 +68,92 @@ conversionHelper(const libMesh::VectorValue<ADReal> & value, const unsigned int
-                             scalar_dof_indices[0],
-                             conversionHelper(residual, i).derivatives()[scalar_dof_indices[0]],
-                             tag);
-        }
```
why did we have all this code ?
were scalar variables not supported with AD at the time? you added it there?
https://github.com/idaholab/moose/pull/13704

lindsayad replied:
Yes I added that support in #13704 and up until now we've been specifically adding Jacobian elements corresponding to local field variable couplings and scalar variable coupling. With the change here to just add all coupling (local *and* nonlocal field variable coupling) then we no longer need to specifically add the scalar contributions

lindsayad committed:
Apply suggestions from code review

- Just return "GhostEverything"
- Add comment describing the use of the average value boundary condition

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad committed:
Don't add component setting parameters for non-vector ADNodalBCs

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1458743/) on 724e69f wanted to post the following:


The following coverage requirement(s) failed:
- `misc` coverage rate 34.31% is less than the required 55.0%
- `phase_field` coverage rate 86.27% is less than the required 89.0%
- `richards` coverage rate 94.75% is less than the required 97.0%
- `tensor_mechanics` coverage rate 85.21% is less than the required 86.0%

moosebuild commented:
Job [Distributed mesh and recover sweep odds](https://civet.inl.gov/job/1461049/) on 977cc43 : invalidated by @lindsayad

moosebuild commented:
Job [Distributed mesh sweep evens](https://civet.inl.gov/job/1461050/) on 977cc43 : invalidated by @lindsayad

moosebuild commented:
Job [Parallel sweep evens](https://civet.inl.gov/job/1461047/) on 977cc43 : invalidated by @lindsayad

moosebuild commented:
Job [Recover sweep odds](https://civet.inl.gov/job/1461045/) on 977cc43 : invalidated by @lindsayad

GiudGiud commented on framework/src/bcs/ADNodalBC.C:
```diff
@@ -44,7 +50,10 @@ ADNodalBCTempl<T>::ADNodalBCTempl(const InputParameters & parameters)
-        {getParam<bool>("set_x_comp"), getParam<bool>("set_y_comp"), getParam<bool>("set_z_comp")})
+        {std::is_same<T, RealVectorValue>::value ? getParam<bool>("set_x_comp") : true,
+         std::is_same<T, RealVectorValue>::value ? getParam<bool>("set_y_comp") : true,
+         std::is_same<T, RealVectorValue>::value ? getParam<bool>("set_z_comp") : true}),
```
I was thinking of straight up removing the parameters. this is fine
